COMP220/COMP220_Portfolio/ForestScene.cpp

```diff
@@ -0,0 +1,215 @@
+
+	glEnable(GL_DEPTH_TEST);
+	glDepthFunc(GL_LESS);
+
+	glEnable(GL_BLEND);
+	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+
+	glEnable(GL_CULL_FACE);
+
+	glm::vec4 playerPosition(0, 0, 5, 1);
+
+	SDL_SetRelativeMouseMode(SDL_TRUE);
+	SDL_GetRelativeMouseState(nullptr, nullptr);
+	Uint32 lastFrameTime = SDL_GetTicks();
+
+	bool running = true;
+	while (running)
+	{
+		SDL_Event ev;
+		while (SDL_PollEvent(&ev))
+		{
+			switch (ev.type)
+			{
+			case SDL_QUIT:
+				running = false;
+				break;
+
+			case SDL_KEYDOWN:
+				switch (ev.key.keysym.sym)
+				{
+				case SDLK_ESCAPE:
+					running = false;
+					break;
+				}
+			}
+		}
+
+		SDL_GetRelativeMouseState(&mouseX, &mouseY);
+		playerYaw -= mouseX * mouseSensitivity;
+		playerPitch -= mouseY * mouseSensitivity;
+		const float maxPitch = glm::radians(89.0f);
+		if (playerPitch > maxPitch)
+			playerPitch = maxPitch;
+		if (playerPitch < -maxPitch)
+			playerPitch = -maxPitch;
+
+		glm::vec4 playerLook(0, 0, -1, 0);
+		glm::mat4 playerRotation;
+		playerRotation = glm::rotate(playerRotation, playerYaw, glm::vec3(0, 1, 0));
+		playerRotation = glm::rotate(playerRotation, playerPitch, glm::vec3(1, 0, 0));
+		playerLook = playerRotation * playerLook;
+
+		glm::vec4 playerForward = playerLook;
+
+		const Uint8* keyboardState = SDL_GetKeyboardState(nullptr);
+		if (keyboardState[SDL_SCANCODE_W])
+		{
+			playerPosition += playerForward * movementMultipler;
+			playerPosition.y = 0;
+		}
+		if (keyboardState[SDL_SCANCODE_S])
+		{
+			playerPosition -= playerForward * movementMultipler;
+			playerPosition.y = 0;
+		}
+
+		glm::vec4 playerRight(0, 0, -1, 0);
+		glm::mat4 playerRightRotation;
+		playerRightRotation = glm::rotate(playerRightRotation, playerYaw - glm::radians(90.0f), glm::vec3(0, 1, 0));
+		playerRight = playerRightRotation * playerRight;
+
+		if (keyboardState[SDL_SCANCODE_A])
+		{
+			playerPosition -= playerRight * movementMultipler;
+			playerPosition.y = 0;
+		}
+		if (keyboardState[SDL_SCANCODE_D])
+		{
+			playerPosition += playerRight * movementMultipler;
+			playerPosition.y = 0;
+		}
+
+		// Calculate delta time
+		Uint32 currentTime = SDL_GetTicks();
+		float deltaTime = (currentTime - lastFrameTime) / 1000.0f;
+		lastFrameTime = currentTime;
+
+		particle.tick(deltaTime);
+
+		glClearColor(0.9f, 0.9f, 0.9f, 1.0f);
+		glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+		glUseProgram(programID);
+
+		glm::mat4 view = glm::lookAt(glm::vec3(playerPosition), glm::vec3(playerPosition + playerLook), glm::vec3(0, 1, 0));
+		glm::mat4 projection = glm::perspective(glm::radians(45.0f), 1.0f, 0.1f, 100.0f);
+
+		glm::mat4 transform;
+		glUniform3f(lightDirectionLocation, 1, 1, 1);
+		glUniform3f(cameraSpaceLocation, playerPosition.x, playerPosition.y, playerPosition.z);
+
+		// Render particles
+		transform = glm::translate(transform, particle.position);
+		transform = glm::rotate(transform, sin(currentTime / 400.0f), glm::vec3(0, 0, 1));
```

Comment:
I don't know, but could these two lines be changed to one line which is "(transform, sin(currentTime / 400.0f), glm::vec3(1, 0, 1))"